pkg/sdn/plugin/egress_dns.go

```diff
@@ -0,0 +1,171 @@
+package plugin
+
+import (
+	"net"
+	"time"
+
+	"github.com/golang/glog"
+
+	osapi "github.com/openshift/origin/pkg/sdn/api"
+
+	kapi "k8s.io/kubernetes/pkg/api"
+	ktypes "k8s.io/kubernetes/pkg/types"
+)
+
+type EgressDNS struct {
+	// Holds Egress DNS entries for each policy
+	pdMap map[ktypes.UID]DNS
+}
+
+func (e *EgressDNS) Add(policy osapi.EgressNetworkPolicy) {
+	dnsInfo := DNS{}
+
+	for _, rule := range policy.Spec.Egress {
+		if len(rule.To.DNSName) > 0 {
+			if err := dnsInfo.Add(rule.To.DNSName); err != nil {
+				glog.Error(err)
+			}
+		}
+	}
+
+	if dnsInfo.Size() > 0 {
+		e.pdMap[policy.UID] = dnsInfo
+	}
+}
+
+func (e *EgressDNS) Delete(policy osapi.EgressNetworkPolicy) {
+	if _, ok := e.pdMap[policy.UID]; ok {
+		delete(e.pdMap, policy.UID)
+	}
+}
+
+func (e *EgressDNS) GetNetCIDRs(policy osapi.EgressNetworkPolicy, dnsName string) []net.IPNet {
+	var cidrs []net.IPNet
+	dnsInfo, ok := e.pdMap[policy.UID]
+	if !ok {
+		return cidrs
+	}
+
+	for name, ips := range dnsInfo.Get() {
+		if name == dnsName {
+			for _, ip := range ips {
+				if ip.To4() != nil { // IPV4 addr
+					cidrs = append(cidrs, net.IPNet{IP: ip, Mask: net.CIDRMask(32, 32)})
+				} else if ip.To16() != nil { // IPV6 addr
+					cidrs = append(cidrs, net.IPNet{IP: ip, Mask: net.CIDRMask(128, 128)})
+				} else {
+					glog.Errorf("Invalid IP: %v for policy: %s", ip, policy.Name)
+				}
+			}
+			break
+		}
+	}
+	return cidrs
+}
+
+func (plugin *OsdnNode) SyncEgressDNSPolicyRules() {
+	t := time.NewTicker(60 * time.Minute)
+	defer t.Stop()
+
+	for {
+		<-t.C
+		glog.V(8).Infof("Periodic egress dns sync: update policy rules")
+
+		policies, err := plugin.osClient.EgressNetworkPolicies(kapi.NamespaceAll).List(kapi.ListOptions{})
+		if err != nil {
+			glog.Errorf("Could not get EgressNetworkPolicies: %v", err)
+			continue
+		}
+
+		for policyUID, dnsInfo := range plugin.egressDNS.pdMap {
+			olddnsMap := dnsInfo.Get()
+			newdnsMap, err := dnsInfo.Update()
+			if err != nil {
+				glog.Error(err)
+				continue
+			}
+			if equal(olddnsMap, newdnsMap) {
+				continue
+			}
+
+			// found dns updates
+			for _, policy := range policies.Items {
+				if policy.UID == policyUID {
+					vnid, err := plugin.policy.GetVNID(policy.Namespace)
+					if err != nil {
+						glog.Warningf("Could not find netid for namespace %q: %v", policy.Namespace, err)
+						break
+					}
+					plugin.updateEgressNetworkPolicyRules(vnid)
```

Comment:
Hm... plugin.updateEgressNetworkPolicyRules() is broken; it needs locking (eg, to protect against concurrent reads from/writes to plugin.egressPolicies[]). And then this will need to use the right locking as well.